### PR TITLE
Working transitions!

### PR DESCRIPTION
**Pull request by daveystruijk**

### CODE REVIEW

File: contextproject-ms3/src/main/java/contextproject/App.java [hunk 1]
```diff
@@ -65,59 +65,63 @@ public static void main(String[] args) {
...
-    Parent root = (Parent) loader.load();
-    final WindowController controller = (WindowController) loader.getController();
-    App.controller = controller;
-
-    Scene scene = new Scene(root, 1200, 800);
-    stage.setTitle("Demo Sprint 3");
-    stage.setScene(scene);
-    stage.show();
-
+  public void start(Stage stage) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/App.java]: please add javadoc comments

File: contextproject-ms3/src/main/java/contextproject/audio/AirhornProcessor.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    
+    try {
+      this.attributes = attributes;
+      this.format = Streamer.streamAudioFormat(attributes);
+    } catch (EncoderException e) {
+      e.printStackTrace();
+    }
+  }
+  
+  public void play() throws EncoderException, LineUnavailableException {
```
Comment (R-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/AirhornProcessor.java:48]: Please add javadoc.

File: contextproject-ms3/src/main/java/contextproject/audio/CustomAudioDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package contextproject.audio;
+
+import be.tarsos.dsp.AudioDispatcher;
+import be.tarsos.dsp.io.TarsosDSPAudioInputStream;
+
+public class CustomAudioDispatcher extends AudioDispatcher {
+
+  public CustomAudioDispatcher(TarsosDSPAudioInputStream stream, int audioBufferSize,
```
Comment (R-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/CustomAudioDispatcher.java:8]: please add javadoc.

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 4]
```diff
@@ -49,70 +67,153 @@ public TrackProcessor(Attributes attributes) {
...
+
+  @Override
+  public boolean process(AudioEvent audioEvent) {
+    currentTime = audioEvent.getTimeStamp();
+    if (transitionTime == 0) {
+      return true;
+    }
+    
+    if (!hasTransitioned && currentTime > transitionTime) {
+      System.out.println(transitionTime);
```
Comment (R-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: if this system.out is needed please add it to a logger, if not please remove

File: contextproject-ms3/src/main/java/contextproject/App.java [hunk 5]
```diff
@@ -65,59 +65,63 @@ public static void main(String[] args) {
...
-          XmlExport exporter = new XmlExport("library.xml", library);
-          exporter.export();
-          System.out.println(library.size());
+      scene.getWindow().setOnHidden(new EventHandler<WindowEvent>() {
+        @Override
+        public void handle(WindowEvent arg0) {
+          if (library != null) {
+            XmlExport exporter = new XmlExport("library.xml", library);
+            exporter.export();
+            System.out.println(library.size());
```
Comment (R-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/App.java]: please remove system.out or replace with a logger

File: contextproject-ms3/src/main/java/contextproject/App.java [hunk 6]
```diff
@@ -65,59 +65,63 @@ public static void main(String[] args) {
...
-      if (selectedDirectory == null) {
-        System.out.println("No directory selected.");
-        System.exit(-1);
+      });
+      if (empty) {
+        String directory = "";
+        DirectoryChooser directoryChooser = new DirectoryChooser();
+        File selectedDirectory = directoryChooser.showDialog(null);
+        if (selectedDirectory == null) {
+          System.out.println("No directory selected.");
```
Comment (R-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/App.java]: same about the system.out here

File: contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java [hunk 7]
```diff
@@ -31,51 +34,67 @@ protected PlayerService() throws EncoderException, LineUnavailableException {
...
   }
+  
   /**
-   * Stops the music.
+   * Starts playing the next track if it's ready, and applies the specified transition.
+   * If the track is not prepared for transition yet, this method will throw an exception.
    */
-  public void exit() {
+  public void setupTransition() {
+    double transitionTime = (60.0 / nextTrack.getBpm()) * 80;
```
Comment (Emiel94) [inline contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java:85]: 80? Make this a local variable. Avoid hardcoding.

File: contextproject-ms3/src/main/java/contextproject/audio/SkipAudioProcessor.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        e.printStackTrace();
+      }
+    }
+    
+    return false;
+  }
+  
+  public void stop() {
+    this.stopped = true;
+  }
```
Comment (Emiel94) [inline contextproject-ms3/src/main/java/contextproject/audio/SkipAudioProcessor.java]: The stop method is not needed anymore.

File: contextproject-ms3/src/main/java/contextproject/audio/transitions/FadeInOutTransition.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+  public FadeInOutTransition(TrackProcessor from, TrackProcessor to, TransitionDoneCallback callback) {
+    super(from, to, callback);
+  }
+
+  @Override
+  public void begin(TrackProcessor from, TrackProcessor to) {
+    to.setGain(0.0);
+    to.play();
+    for (int i = 0; i < 16; i++) {
```
Comment (Emiel94) [inline contextproject-ms3/src/main/java/contextproject/audio/transitions/FadeInOutTransition.java:15]: Local varaible instead of 16. Avoid hardcoding

**Pull request merged**